src/mpfao/3D/tdympfao3D_core.c

```diff
@@ -979,3 +1032,317 @@ PetscErrorCode TDyComputeTransmissibilityMatrix3DMesh(TDy tdy) {
+/* -------------------------------------------------------------------------- */
+/// Computes face permeability tensor as a harmonically distance-weighted
+//  permeability of upwind and downwind permeability tensors
+///
+/// K = (wt_1 * K_u^{-1} + (1-wt_1) * K_d^{-1})^{-1}
+///
+/// @param [in] tdy A TDy struct
+/// @param [in] face_id ID of the face
+/// @param [out] Kface[9] face permeability tensor
+PetscErrorCode ComputeFacePermeabilityTensor(TDy tdy, PetscInt face_id, PetscReal Kface[9]){
+
+  PetscFunctionBegin;
+
+  PetscErrorCode ierr;
+  PetscReal dist_up, dist_dn;
+
+  ierr = ComputeUpAndDownDist(tdy, face_id, &dist_up, &dist_dn); CHKERRQ(ierr);
+
+  PetscInt dim = 3;
+
+  PetscReal Kup[dim*dim], Kdn[dim*dim];
+  PetscReal KupInv[dim*dim], KdnInv[dim*dim];
+
+  ierr = ExtractUpAndDownPermeabilityTensors(tdy, face_id, dim, Kup, Kdn); CHKERRQ(ierr);
+
+  ierr = ComputeInverseOf3by3Matrix(Kup, KupInv); CHKERRQ(ierr);
+  ierr = ComputeInverseOf3by3Matrix(Kdn, KdnInv); CHKERRQ(ierr);
+
+  PetscReal KfaceInv[dim*dim];
+  PetscReal wt_up = dist_up / (dist_up + dist_dn);
+
+  PetscInt idx;
+
+  // Compute (wt_1 * K_u^{-1} + (1-wt_1) * K_d^{-1})
+  idx = 0;
+  for (PetscInt kk = 0; kk < dim; kk++) {
+    for (PetscInt mm = 0; mm < dim; mm++) {
+      KfaceInv[idx] = wt_up * KupInv[idx] + (1.0 - wt_up) * KdnInv[idx];
+      idx++;
+    }
+  }
+
+  ierr = ComputeInverseOf3by3Matrix(KfaceInv, Kface); CHKERRQ(ierr);
+
+  idx = 0;
+  for (PetscInt kk = 0; kk < dim; kk++) {
+    for (PetscInt mm = 0; mm < dim; mm++) {
+      if (Kface[idx] < 0.0 || fabs(Kface[idx] < PETSC_MACHINE_EPSILON) ) Kface[idx] = 0.0;
+      idx++;
+    }
+  }
+
+ PetscFunctionReturn(0);
+}
+
+/* -------------------------------------------------------------------------- */
+/// Computes the two-point flux discretization of the gravity term in Richards equation
+///
+/// Flux associated with the gravity term is given as
+///
+/// flux_m = rho^2 * ukvr * A_face * dot (n_face, K_face x u_up2dn) * dot(g, u_up2dn)
+///        = rho^2 * ukvr * GravDis
+///
+/// where
+///   rho     = distance weighted density
+///   ukvr    = upwind relative permeability divided by viscosity
+///   A_face  = face area
+///   n_face  = normal to face area
+///   K_face  = permeability tensor at the face
+///   u_up2dn = unit vector from upwind to downwind cell connected through a common face
+///   g       = gravity vector
+///   GravDis = gravity discretization term that is not dependent on the unknown variable(s)
+///             such as pressure, temperautre. Thus, this term is precomputed.
+/// 
+/// Starnoni, M., Berre, I., Keilegavlen, E., & Nordbotten, J. M. (2019).
+/// Consistent mpfa discretization for flow in the presence of gravity. Water
+/// Resources Research, 55, 10105– 10118. https://doi.org/10.1029/2019WR025384
+/// 
+/// @param [in] tdy A TDy struct
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode TDyComputeGravityDiscretizationFor3DMesh(TDy tdy) {
```

Comment:
We could call this something like `TDyComputeTPFGravityTerm3D` to indicate that it's a TPF term.

Reply:
In a future PR, this function will be used by the non-TPF code.